### ISSUE TITLE
(doc) \cs_generate_variant:Nn needs a defined parent function?

### ISSUE THREAD

**Issue by Vidabe** [latex3/latex3#983]
Quote from the documentation:

> The \cs_generate_variant:Nn function can only be applied if the ⟨parent control sequence⟩ is already defined.

But as the following minimal example shows, the parent function does not need to be defined to be used
in `\cs_generate_variant:Nn`:
```
\documentclass{article}
\begin{document}
\ExplSyntaxOn
\cs_if_exist:NTF \__I_am_not_defined:Nn { Defined } { Nope }
\cs_generate_variant:Nn \__I_am_not_defined:Nn { cn }
\ExplSyntaxOff
\end{document}
```

Using `\__I_am_not_defined:cn` of created an `Undefined control sequence: \__I_am_not_defined:Nn` error.

**Comment by FrankMittelbach** [comment]
As it stands it is a white lie. But not quite: if you enable the debugging option "check-declarations" it is tested and reported. In general we don't do all kind of runtime tests for programming errors but instead do them only if explicitly asked for.
This is similar to the statement that `\g_`variables should only be used with `\foo_g...` functions. It produces chaos if you don't, but we don't test at runtime (normally).

```
\documentclass{article}

\usepackage[enable-debug]{expl3}  % this will soon not be necessary to use debug_on
\begin{document}
\ExplSyntaxOn
\debug_on:n { check-declarations }
\cs_if_exist:NTF \__I_am_not_defined:Nn { Defined } { Nope }
\cs_generate_variant:Nn \__I_am_not_defined:Nn { cn }
\ExplSyntaxOff
\end{document}
```
This then gives you:
```
\documentclass{article}

\usepackage[enable-debug]{expl3}  % this will soon not be necessary to use debug_on
\begin{document}
\ExplSyntaxOn
\debug_on:n { check-declarations }
\cs_if_exist:NTF \__I_am_not_defined:Nn { Defined } { Nope }
\cs_generate_variant:Nn \__I_am_not_defined:Nn { cn }
\ExplSyntaxOff
\end{document}
```

Maybe at the beginning of the reference manual that point should be stressed and also how (as a developer you can and should have a test file to ensure that your code is clean (while still being fast in production)

**Comment by Vidabe** [comment]
I guess this is similar to #972, although I am not quite sure if I understand the cost-efficiency argument for `\cs_generate_variant:Nn`. I understand it for `prop_item:cn` as it can be used in an expansion or mathematical expression and thus checking its existence everytime is redundant. But I do not really see `\cs_generate_variant:Nn` as something that is used in a loop or in an expression, but rather as an addendum to `\cs_new:Npn` (So the number of `\cs_generate_variant:Nn` in a document is capped by the number of`\cs_new:Npn`). Would it really slow things down that much if it checks its first argument for existence? 

Just some thoughts and curiosity on that matter. I also did not know that declaration-checking is costly in LaTeX (I always assumed it to be a quick and easy check).

**Comment by josephwright** [comment]
I've tightened up the docs.

**Comment by muzimuzhi** [comment]
(resolved in ae2556dc39871e735920772eb89c0322aa648cfb)
